### DISCUSSION TITLE
How to deal with derived values from nested states and avoid re-rendering?

### DISCUSSION THREAD

**Discussion by bruno-teruel** [pmndrs/zustand#2406] [Q&A]
Hello, 

I am new to zustand and I have a question on getting derived state from a nested state.

I have a nested state that is made up of a dicitonary with UUIDs, some of the values in the dict are the objects themselves, others are a list of objects, which contains arrays to UUIDs. For example:
```
{{key: root, group: [3]},
{key: 1, properties: ...},
{key: 2, properties: ...},
{key: 3, group: [1, 2]}}
```

After reading this thread: [132](https://github.com/pmndrs/zustand/issues/132), I tried to follow the patterns recommended there and eventually got to this, which has been working fairly well. However, my issue is that the state changesin one property cause a rerender for every component that calls my getters, even if the property used in the getter is not changed, and that results in unecessary re-renders and performance issues. I feel like I am using Zustand incorrectly. What would be the recommended way of dealing with derived values from nested states?

```
export const useBearStore = create()(set) => ({
    bears: {{key: 1, value: 55}, {key: 2, value: 43}, {key: 3, group: [1, 2]}}
})

export bearGetters = {
    getBearValue: (state) => (id) => {
          if (isGroup(state.bears[id)) {
                return Math.sum(state.bears[id].map((id) =>getBearValue(state)(id)))
         }
         else {
                return state.bears[id].value
         }
    }
}

const Component = (props) => {
         const getBearValue = useBearStore((state) => bearGetters.getBearValue(state))
         ...
```

**Comment by Kamilcat** [comment]
@bruno-teruel you have to use `useShallow` to prevent re-renders on derived values and nested objects:

```
import { useShallow } from 'zustand/react/shallow';

// ...

const getBearValue = useBearStore(useShallow((state) => bearGetters.getBearValue(state)))

```
